### PR TITLE
C++: Make realloc a data-flow function

### PR DESCRIPTION
**Pull request by paldepind**
Make `realloc` a data-flow function instead of a taint function.

### CODE REVIEW

File: cpp/ql/test/library-tests/dataflow/taint-tests/taint.cpp [hunk 1]
```diff
@@ -746,6 +747,14 @@ void test_realloc_2_indirections(int **buffer) {
   sink(**buffer); // $ ir MISSING: ast
 }
 
+void test_realloc_struct_field() {
+	struct A { int x; };
+	A* a = (A*)malloc(sizeof(A));
+	a->x = source();
+	A* a2 = (A*)realloc(a, sizeof(A));
+	sink(a2->x); // $ ir MISSING: ast
+}
```
Comment (jketema) [inline cpp/ql/test/library-tests/dataflow/taint-tests/taint.cpp:756]: I would have expected a dataflow test, not a taint test. Presumably this should already have worked without any changes to the library?

Comment (paldepind) [reply cpp/ql/test/library-tests/dataflow/taint-tests/taint.cpp:756]: Agree, I would also have expected a data-flow test :) But this test is what the internal issue proposed. I think one could say that this test is more of an integration test, ensuring that the added data-flow integrates with taint tracking to give the desired outcome.

Prior to this PR both `ir` and `ast` would be missing on L755. As far as I understand, this is because `realloc` being a taint step does not propagate precise information about it's fields, but this happens now since the data-flow step preserves all the field tracking for `a`.

Comment (jketema) [reply cpp/ql/test/library-tests/dataflow/taint-tests/taint.cpp:756]: Still I expect this to also work as a data flow test, or will that have missing results for IR?

Comment (paldepind) [reply cpp/ql/test/library-tests/dataflow/taint-tests/taint.cpp:756]: Sure, I've added a data-flow test as well 👍

File: cpp/ql/test/library-tests/dataflow/dataflow-tests/test.cpp [hunk 2]
```diff
@@ -1046,6 +1046,14 @@ void memset_test(char* buf) { // $ ast-def=buf ir-def=*buf
 	sink(*buf); // $ ir MISSING: ast
 }
 
+void *realloc(void *, size_t);
+
+void test_realloc() {
+	int *src = indirect_source();
+	int *dest = (int*)realloc(src, sizeof(int));
+	sink(*dest); // $ ir, MISSING: ast
```
Comment (paldepind) [inline cpp/ql/test/library-tests/dataflow/dataflow-tests/test.cpp:1054]: Note that `ast` is missing. The same is true for the similarly modeled `memset` just above, so I think this is expected.

Comment (jketema) [reply cpp/ql/test/library-tests/dataflow/dataflow-tests/test.cpp:1054]: I'm looking into removing the AST dataflow library.

File: cpp/ql/test/library-tests/dataflow/taint-tests/taint.cpp (hunk 1 above)
Comment (jketema) [reply cpp/ql/test/library-tests/dataflow/taint-tests/taint.cpp:756]: I think some of the expected test results need to be updated.

**Comment by jketema** [comment]
This looks sensible to me. I do think we need to run DCA.

File: cpp/ql/test/library-tests/dataflow/dataflow-tests/test.cpp (hunk 2 above)
Comment (MathiasVP) [inline cpp/ql/test/library-tests/dataflow/dataflow-tests/test.cpp:1054]: FWIW, this is not really testing that `realloc` is a `dataflow` function. This test would also pass with `realloc` being a taint-function.

See the testcase I provided in the internal issue if you want to fix this 😄 I don't really mind, though.

File: cpp/ql/test/library-tests/dataflow/taint-tests/taint.cpp (hunk 1 above)
Comment (MathiasVP) [reply cpp/ql/test/library-tests/dataflow/taint-tests/taint.cpp:756]: > Presumably this should already have worked without any changes to the library?

No it wouldn't. Unlike dataflow functions, taint functions aren't taken into account when the dataflow library is tracking a non-empty access path. That's why I added the field write/read in the test in the issue.

Likewise, dataflow functions are also used when computing flow summaries. So we could also have created a testcase based on this.

File: cpp/ql/test/library-tests/dataflow/dataflow-tests/test.cpp [hunk 3]
```diff
@@ -1046,6 +1046,14 @@ void memset_test(char* buf) { // $ ast-def=buf ir-def=*buf
...
+}
```
Comment (jketema) [inline cpp/ql/test/library-tests/dataflow/dataflow-tests/test.cpp:1055]: I expected the same test here as for taint tracking. Why did you add a different test?

Comment (paldepind) [reply cpp/ql/test/library-tests/dataflow/dataflow-tests/test.cpp:1055]: For the other test having a struct with a field is necessary to tease out the effect that the change in this PR has on taint tracking (or it is at least one way to do so). That complexity is not needed here, where this simpler test is enough to see that `realloc` carries data-flow through it.

The [test query](https://github.com/paldepind/codeql/blob/d5948d2d95f9b0f25416cfcf8428487533622e93/cpp/ql/test/library-tests/dataflow/dataflow-tests/test.ql#L3) uses [`AstFlow`](https://github.com/paldepind/codeql/blob/d5948d2d95f9b0f25416cfcf8428487533622e93/cpp/ql/test/library-tests/dataflow/dataflow-tests/TestBase.qll#L51) which only uses data-flow. So if `realloc` had not been modeled as a data-flow function the path would have been cut off at L1054.

File: cpp/ql/lib/change-notes/2024-09-03-realloc-data-flow.md [hunk 4]
```diff
@@ -0,0 +1,4 @@
+---
+category: minorAnalysis
+---
+* Added a data-flow model for `realloc` and related functions. Previously they was modeled as a taint functions. This improves the precision of queries using data-flow or taint-tracking when `realloc` is involved.
```
Comment (jketema) [inline cpp/ql/lib/change-notes/2024-09-03-realloc-data-flow.md:4]: Fixing some of the grammar, and trying to make it slightly clearer.
```suggestion
* Added a data flow model for `realloc`-like functions, which were previously modeled as a taint tracking functions. This change improves the precision of queries where flow through `realloc`-like functions might affect the results.
```

File: cpp/ql/test/library-tests/dataflow/dataflow-tests/test.cpp (hunk 3 above)
Comment (jketema) [reply cpp/ql/test/library-tests/dataflow/dataflow-tests/test.cpp:1055]: > The [test query](https://github.com/paldepind/codeql/blob/d5948d2d95f9b0f25416cfcf8428487533622e93/cpp/ql/test/library-tests/dataflow/dataflow-tests/test.ql#L3) uses [`AstFlow`](https://github.com/paldepind/codeql/blob/d5948d2d95f9b0f25416cfcf8428487533622e93/cpp/ql/test/library-tests/dataflow/dataflow-tests/TestBase.qll#L51) which only uses data-flow. So if `realloc` had not been modeled as a data-flow function the path would have been cut off at L1054.

Ast and IR data flow, to be exact 😄

**Review by jketema — APPROVED** [review]

**Comment by paldepind** [comment]
Thanks for reviewing @jketema and thanks to @MathiasVP for chiming in 👍
